syntax = "proto3";

package aea.eightballer.chatroom.v0_1_0;

message ChatroomMessage{

  // Custom Types
  message ErrorCode{
    enum ErrorCodeEnum {
      ERROR_CODE_ENUM_UNKNOWN_CHAT_ID = 0;
      ERROR_CODE_ENUM_API_ERROR = 1;
      ERROR_CODE_ENUM_INVALID_MESSAGE_FORMAT = 2;
    }
    ErrorCodeEnum error_code = 1;
  }


  // Performatives and contents
  message Message_Performative{
    string chat_id = 1;
    string text = 2;
    int32 id = 3;
    bool id_is_set = 4;
    string parse_mode = 5;
    bool parse_mode_is_set = 6;
    string reply_markup = 7;
    bool reply_markup_is_set = 8;
    string from_user = 9;
    bool from_user_is_set = 10;
    int32 timestamp = 11;
    bool timestamp_is_set = 12;
  }

  message Message_Sent_Performative{
    int32 id = 1;
    bool id_is_set = 2;
  }

  message Error_Performative{
    ErrorCode error_code = 1;
    string error_msg = 2;
    map<string, bytes> error_data = 3;
  }

  message Subscribe_Performative{
    string chat_id = 1;
  }

  message Unsubscribe_Performative{
    string chat_id = 1;
  }

  message Get_Channels_Performative{
    string agent_id = 1;
  }

  message Unsubscription_Result_Performative{
    string chat_id = 1;
    string status = 2;
  }

  message Subscription_Result_Performative{
    string chat_id = 1;
    string status = 2;
  }

  message Channels_Performative{
    repeated string channels = 1;
  }


  oneof performative{
    Channels_Performative channels = 5;
    Error_Performative error = 6;
    Get_Channels_Performative get_channels = 7;
    Message_Performative message = 8;
    Message_Sent_Performative message_sent = 9;
    Subscribe_Performative subscribe = 10;
    Subscription_Result_Performative subscription_result = 11;
    Unsubscribe_Performative unsubscribe = 12;
    Unsubscription_Result_Performative unsubscription_result = 13;
  }
}
